Source/ARTRealtimeChannel.m

```diff
@@ -727,6 +729,10 @@ - (void)setAttached:(ARTProtocolMessage *)message {
     }
 
     self.attachSerial = message.channelSerial;
+    // RTL15b
+    if (message.channelSerial) {
+        self.serial = message.channelSerial;
+    }
 
```

Comment:
There should be a code right below L735 =>
```
    if (self.state_nosync == ARTRealtimeChannelAttached) {
        if (message.resumed) { // RTL12
        return;
        } else {
            if (message.error != nil) {
                _errorReason = message.error;
            }
            ARTChannelStateChange *stateChange = [[ARTChannelStateChange alloc] initWithCurrent:self.state_nosync previous:self.state_nosync event:ARTChannelEventUpdate reason:message.error resumed:message.resumed];
            [self emit:stateChange.event with:stateChange];
        }
    }
```

Reply:
Block of code from L749-757 should be removed.

Reply:
Also, please note that `message.resumed` checks if message has `HAS_PRESENCE` flag. It is bitwise flag with value 
`1 << 0`.
In case of `message.Flags.Has(flagResumed)` we are checking for a bit flag of value `1 << 2`

Reply:
You can check all of those flags ->
https://github.com/ably/ably-go/blob/cc9e260bf265f89cb42fd8f19b62b904fef2d1c8/ably/proto_protocol_message.go#L9-L19

Reply:
>There should be a code right below L735

How is it in line with [(RTP17f)](https://sdk.ably.com/builds/ably/specification/main/features/#RTP17f):
_"The RealtimePresence object should perform automatic re-entry whenever a channel moves into the ATTACHED state. (It does not need to do so for RTL12 ATTACHED events received on already-ATTACHED channels)"_.

And [(RTL12)](https://sdk.ably.com/builds/ably/specification/main/features/#RTL12):
_"...Note that UPDATE should be the only event emitted: in particular, the library must not emit an ATTACHED event if the channel was already attached..."_ ?

>Also, please note that message.resumed checks if message has HAS_PRESENCE flag. It is bitwise flag with value

Not sure why you say so, [`resumed`](https://github.com/ably/ably-cocoa/blob/d1691e6f73ac93d92dc456bdfe1329e0d2846f29/Source/ARTProtocolMessage.m#L174) checks `ARTProtocolMessageFlagResumed` which is `(1UL << 2)`
https://github.com/ably/ably-cocoa/blob/d1691e6f73ac93d92dc456bdfe1329e0d2846f29/Source/ARTProtocolMessage.m#L175

Reply:
Yeah, my bad. Seems we are using right flag. Though we would still need to implement the same code
```
    if (self.state_nosync == ARTRealtimeChannelAttached) {
        if (message.resumed) { // RTL12
        return;
        } else {
            if (message.error != nil) {
                _errorReason = message.error;
            }
            ARTChannelStateChange *stateChange = [[ARTChannelStateChange alloc] initWithCurrent:self.state_nosync previous:self.state_nosync event:ARTChannelEventUpdate reason:message.error resumed:message.resumed];
            [self emit:stateChange.event with:stateChange];
        }
    }
```

Reply:
This would remove duplicated use of `if (!message.resumed)` in the underlying code.

Reply:
https://github.com/ably/ably-cocoa/pull/1714/commits/410948379e3ee6b187a293d0da39671e12ba78e8

Reply:
About code
```
    else if ([self.presenceMap.members count] > 0 || [self.presenceMap.localMembers count] > 0) {
        if (!message.resumed) {
            // When an ATTACHED message is received without a HAS_PRESENCE flag and PresenceMap has existing members
            [self.presenceMap startSync];
            [self.presenceMap endSync];
            ARTLogDebug(self.logger, @"R:%p C:%p (%@) PresenceMap has been reset", _realtime, self, self.name);
        }
    }
```
We don't need check `if (!message.resumed)`

Reply:
Code can be simplified as -> 
    else {
            [self.presenceMap startSync];
            [self.presenceMap endSync];
            ARTLogDebug(self.logger, @"R:%p C:%p (%@) PresenceMap has been reset", _realtime, self, self.name);
        }
Since, we are anyways iterating over members, for zero size array, it won't execute any operations as such

Reply:
https://github.com/ably/ably-cocoa/pull/1714/commits/646ee4b5d1c4d047cc9cdcfe5c6d75ed83f18b05

Reply:
https://github.com/ably/ably-cocoa/pull/1869

Reply:
Lgtm

Reply:
I've unresolved this conversation because I don't understand the change introduced in 410948379e3ee6b187a293d0da39671e12ba78e8 (why does the RTP17f behaviour change if the `resumed` flag is not set?), and the link that you've added doesn't seem to be of any use — it just takes me to the diff for the PR.

Reply:
Why did we decide to do this without updating the spec?

Reply:
I thought we had previously agreed we were going to decide the correct behaviour and then update the spec.

Reply:
https://github.com/ably/ably-cocoa/pull/1714/commits/c7da04b7f9ab6a396b95043c431b49bf1487636f (and https://github.com/ably/ably-cocoa/pull/1714/commits/52313f2de0042e8569f74d382b99d0715d417fb6)

Reply:
To avoid a bunch of churn on this PR, I suggest not making any further changes until that spec PR is approved and merged.

Reply:
The tests still refer to RTP17f.

Reply:
https://github.com/ably/ably-cocoa/pull/1714/commits/fb01ab84831d3926262851393820296262a38349

Reply:
https://github.com/ably/ably-cocoa/pull/1714/commits/6e62c706b22fcc57cca11587b7bdf8907d9da19f